### ISSUE TITLE
unnamed register temporarily broken after writing into "+ from another Vim instance

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6435]
**Describe the bug**

If the unnamed register points to `"+`, it gets temporarily broken after writing into the `"+` register from another Vim instance.

**To Reproduce**

From a terminal, run this shell command:

    vim -Nu NONE -i NONE +"pu!='first' | norm! \"+y$"

Without closing the first terminal, run this shell command from another terminal:

    vim -Nu NONE -i NONE +'let @+ = "second"'

Focus back the first Vim instance, and run this Ex command:

    :reg "

It outputs:

    Type Name Content

**Expected behavior**

It outputs:

    Type Name Content
      c  ""   second

**Environment**

 - Vim version: 8.2 Included patches: 1-1182
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(356)

**Additional context**

For some reason, writing into `"+` from a second Vim instance breaks the connection from `""` to `"+` in the first one.

---

If you run `:reg "` a second time, then the output is correct:

    Type Name Content
      c  ""   second

---

This sometimes causes an issue when I use the `p` or `P` command.  I have a mapping and an autocmd which edit the `"+` register before pasting it.  If one of them is used in a Vim instance, while `""` points to `"+` in another one, then the connection is lost in the latter; until I run `:reg "`.

To fix this, I need these extra mappings whose only purpose is to run `:reg "` whenever I try to put some text:

    nno <expr> p <sid>put('p')
    nno <expr> P <sid>put('P')
    xno <expr> p <sid>put('p')
    xno <expr> P <sid>put('P')

    fu s:put(key) abort
        sil reg "
        return a:key
    endfu

---

I don't know whether this is intended or a bug; if it is a bug, then it's caused by a regression introduced in dd82d699c8cb3e60b2da968fb40622dc497f71ca.

**Comment by brammool** [comment]
I'm not quite sure what the correct behavior is.  At least it is strange when a regular Vim register gets changed by an action outside of this Vim instance.  That the "+ register changes is to be expected, but that "" changes is not.
When yanking into the "+ register, that's the only place where the text is stored.  Once the ownership of the clipboard is taken over by another process, then the text is gone.  We could have the "" register follow "+ still, but it's a bit strange.
The only other solution I can think of is when ownership of "+ is lost then a copy of the text is made for the "" register.  That's a bit strange, since "" normally refers to another register, does not contain text by itself.

**Comment by lacygoill** [comment]
> We could have the "" register follow "+ still, but it's a bit strange.

Yes, that's what I think would be the least unexpected behavior.

Note that Vim is already doing this right now; only for some reason, you have to ask it what is the contents of the unnamed register via `:reg "` for `""` to follow `"+` again, which in itself is also a bit strange; I would not expect `:reg` to have any side effect aside from printing some text.
